clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v1/deploy/ops/servergroup/DeployKubernetesAtomicOperation.groovy

```diff
@@ -170,35 +172,76 @@ class DeployKubernetesAtomicOperation implements AtomicOperation<DeploymentResul
+    }
 
-      task.updateStatus BASE_PHASE, "Deployed stateful set ${controllerSet.metadata.name}"
-      controllerSet = credentials.clientApiAdaptor.createStatfulSet(namespace, controllerSet)
+    return KubernetesClientApiConverter.toKubernetesController(controllerSet)
+  }
 
-      if (description.scalingPolicy) {
-        task.updateStatus BASE_PHASE, "Attaching a horizontal pod autoscaler..."
+  def deployStatefulSet(KubernetesV1Credentials credentials, String controllerName, String clusterName, String namespace,  Boolean canUpdated) {
+    task.updateStatus BASE_PHASE, "Building stateful set..."
+    def controllerSet = KubernetesClientApiConverter.toStatefulSet(description, controllerName)
+    if (canUpdated) {
+      def deployedControllerSet = credentials.clientApiAdaptor.getStatefulSet(controllerName, namespace)
+      if (deployedControllerSet) {
+        task.updateStatus BASE_PHASE, "Update stateful set ${controllerName}"
+        controllerSet = credentials.clientApiAdaptor.replaceStatfulSet(controllerName, namespace, controllerSet)
+      } else {
+        task.updateStatus BASE_PHASE, "Deployed stateful set ${controllerName}"
+        controllerSet = credentials.clientApiAdaptor.createStatfulSet(namespace, controllerSet)
+      }
+    } else {
+      task.updateStatus BASE_PHASE, "Deployed stateful set ${controllerName}"
+      controllerSet = credentials.clientApiAdaptor.createStatfulSet(namespace, controllerSet)
+    }
 
-        def autoscaler = KubernetesClientApiConverter.toAutoscaler(new KubernetesAutoscalerDescription(controllerName, description), controllerName, description.kind)
+    if (description.scalingPolicy) {
+      task.updateStatus BASE_PHASE, "Attaching a horizontal pod autoscaler..."
 
-        if (credentials.clientApiAdaptor.getAutoscaler(namespace, controllerName)) {
-          credentials.clientApiAdaptor.deleteAutoscaler(namespace, controllerName)
-        }
+      def autoscaler = KubernetesClientApiConverter.toAutoscaler(new KubernetesAutoscalerDescription(controllerName, description), controllerName, description.kind)
 
-        credentials.clientApiAdaptor.createAutoscaler(namespace, autoscaler)
+      if (credentials.clientApiAdaptor.getAutoscaler(namespace, controllerName)) {
+        credentials.clientApiAdaptor.deleteAutoscaler(namespace, controllerName, null, null, null)
       }
+      credentials.clientApiAdaptor.createAutoscaler(namespace, autoscaler)
     }
-    else if (description.kind == KubernetesUtil.CONTROLLERS_DAEMONSET_KIND) {
-      task.updateStatus BASE_PHASE, "Building daemonset set..."
-      controllerSet = KubernetesClientApiConverter.toDaemonSet(description, controllerName)
 
-      task.updateStatus BASE_PHASE, "Deployed daemonset set ${controllerSet.metadata.name}"
+    return controllerSet
+  }
+
+  def deployDaemonSet(KubernetesV1Credentials credentials, String controllerName, String clusterName, String namespace, boolean canUpdated) {
```

Comment:
What is this doing? There is not daemon set in the request

Reply:
I would move the description into the method to make explicit what this is deploying

Reply:
deploydemonset is for switch statement.